akd/src/directory.rs

```diff
@@ -276,24 +303,42 @@ impl<S: Storage + Sync + Send> Directory<S> {
     }
 
     /// Poll for changes in the epoch number of the AZKS struct
-    /// stored in the storage layer. If an epoch change is detected, the cache is flushed immediately
-    /// so that new objects are retrieved from the storage layer
+    /// stored in the storage layer. If an epoch change is detected,
+    /// the object cache (if present) is flushed immediately so
+    /// that new objects are retrieved from the storage layer against
+    /// the "latest" epoch. There is a "special" flow in the storage layer
+    /// to do a storage-layer retrieval which ignores the cache
     pub async fn poll_for_azks_changes(
         &self,
         period: tokio::time::Duration,
+        change_detected: Option<tokio::sync::mpsc::Sender<()>>,
     ) -> Result<(), AkdError> {
-        let mut last = Directory::get_azks_from_storage(&self.storage, true).await?;
+        // Retrieve the same AZKS that all the other calls see (i.e. the version that could be cached
+        // at this point). We'll compare this via an uncached call when a change is notified
+        let mut last = Directory::get_azks_from_storage(&self.storage, false).await?;
 
         loop {
             // loop forever polling for changes
             tokio::time::sleep(period).await;
 
             let latest = Directory::get_azks_from_storage(&self.storage, true).await?;
             if latest.latest_epoch > last.latest_epoch {
-                // an update was detected
-                last = latest;
-
-                self.storage.flush_cache().await;
+                {
+                    // acquire a singleton lock prior to flushing the cache to assert that no
+                    // cache accesses are underway (i.e. publish/proof generations/etc)
+                    let _guard = self.cache_lock.write().await;
+                    // flush the cache in its entirety
+                    self.storage.flush_cache().await;
+                    // re-fetch the azks to load it into cache so when we release the cache lock
+                    // others will see the new AZKS loaded up and ready
+                    last = Directory::get_azks_from_storage(&self.storage, false).await?;
+
+                    // notify change occurred
+                    if let Some(channel) = &change_detected {
+                        channel.send(()).await.map_err(|send_err| AkdError::Directory(DirectoryError::Storage(StorageError::Connection(format!("Tokio MPSC sender failed to publish notification with error {:?}", send_err)))))?;
```

Comment:
nit: multiple lines for readability

Reply:
Unfortunately this is rustfmt. I could extract to a function or wrap in curly braces, but Rust chose the linewrapping here